core/src/oop_module.c

```diff
@@ -0,0 +1,720 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+#include "oop_module.h"
+#include "broker.h"
+
+#include "stdlib.h"
+#include "string.h"
+
+#include "nanomsg/nn.h"
+#include "nanomsg/pair.h"
+
+#include "azure_c_shared_utility/gballoc.h"
+#include "azure_c_shared_utility/lock.h"
+#include "azure_c_shared_utility/threadapi.h"
+#include "azure_c_shared_utility/xlogging.h"
+
+#include "control_message.h"
+#include "gateway.h"
+#include "message.h"
+
+typedef enum OOP_MODULE_STATUS_TAG {
+    OOP_MODULE_DETACH = -1,
+    OOP_MODULE_OK,
+    OOP_MODULE_GATEWAY_CONNECTION_ERROR,
+    OOP_MODULE_MODULE_CREATION_ERROR,
+} OOP_MODULE_STATUS;
+
+typedef struct MESSAGE_THREAD_TAG * MESSAGE_THREAD_HANDLE;
+
+int
+connect_to_message_channels (
+    OOP_MODULE_HANDLE oop_module,
+    const MESSAGE_URI * channel_uris,
+    const size_t channel_count
+);
+
+void
+disconnect_from_message_channels (
+    OOP_MODULE_HANDLE oop_module
+);
+
+int
+invoke_add_module_process (
+    OOP_MODULE_HANDLE oop_module,
+    const char * json_config
+);
+
+int
+worker_thread (
+    void * thread_arg
+);
+
+int
+process_module_create_message (
```

Comment:
I see the word "process" used in this file two different ways:
1. Out-of-**process** module (an instance of a program)
2. `process_module_create_message` and `invoke_add_module_process` (set of activities that achieve a result)

For (2) can we call it something else to avoid confusion? Something like `handle_module_create_message` and `init_module`...